### PR TITLE
Disallow dangerous casts where possible

### LINKED ISSUE #958: Enable clippy lints about dangerous casts
Recently there was an overflow bug with HTTP errors in radon, where the status code was casted from `u16` to `u8`. To prevent this kind of bugs in the future, we can enable additional clippy lints about dangerous numeric casts.

Add clippy lints to [Justfile](https://github.com/witnet/witnet-rust/blob/e0c8e6a2fef7375383155fd86c70a0b49ccd3861/Justfile#L29):

```
# additional clippy lints
export CLIPPY_LINTS := '-D warnings
    -D clippy::cast-lossless
    -D clippy::cast-possible-truncation
    -D clippy::cast-possible-wrap
    -D clippy::cast-precision-loss
    -D clippy::cast-sign-loss
    -D clippy::checked-conversions
'
# run clippy
clippy +flags="":
    cargo clippy --all --all-features -- $CLIPPY_LINTS {{flags}}
    cargo clippy --all --all-targets --all-features -- $CLIPPY_LINTS {{flags}}
```

[More details about relevant clippy lints](https://rust-lang.github.io/rust-clippy/master/index.html#cast)

The main blocker is that we use hundreds of casts like `x_i64 as u64` and `x_u64 as usize`. But most of the time there shouldn't be a problem with replacing it with `u64::try_from(x_i64).unwrap()`. It's better to panic than to silently truncate the value.

When casting to a greater value, it's more idiomatic to use `u64::from(x_u32)` than `x_u32 as u64`, because if the type of x changes to `i64` for example, the first will stop compiling while the second will silently cast `-1` to `u64::max_value()`.

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #958.

### CODE REVIEW

File: rad/src/operators/array.rs [hunk 1]
```diff
@@ -49,14 +49,16 @@ pub fn get(input: &RadonArray, args: &[Value]) -> Result<RadonTypes, RadError> {
...
     }
 
-    let not_found = |index: i32| RadError::ArrayIndexNotFound { index };
+    let not_found = |index: usize| RadError::ArrayIndexNotFound {
+        index: i32::try_from(index).unwrap(),
+    };
 
     let arg = args[0].to_owned();
-    let index = from_value::<i32>(arg).map_err(|_| wrong_args())?;
+    let index = from_value::<usize>(arg).map_err(|_| wrong_args())?;
```
Comment (tmpolaczyk) [inline rad/src/operators/array.rs]: Good find, it doesn't make sense to index an array with negative values. This has some problems, I opened issue #1087 in order to further discuss that. In the meanwhile we can just return `not_found` error on negative values:

```suggestion
    let index = from_value::<i32>(arg).map_err(|_| wrong_args())?;
    let index = usize::try_from(index).map_err(|_| not_found(index))?;
```

**Comment by tmpolaczyk** [comment]
Wow, thanks!

I left some comments but overall this PR is really good.

We will have to assess the benefits of having these two lints, as it looks like they are too restrictive.

    // FIXME: Allow for now, wait for https://github.com/rust-lang/rust/issues/67058 to reach stable
    #[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]

Maybe we can just not enable those lints, or disable them for tests? They look useful but if they don't work with float conversions they will just make the code less readable with no benefits...

File: rad/src/script.rs [hunk 2]
```diff
@@ -72,7 +72,9 @@ pub fn unpack_radon_call(packed_call: &Value) -> Result<RadonCall, RadError> {
         Value::Array(array) => unpack_compound_call(array),
         Value::Integer(integer) => {
             if *integer >= 0i128 {
-                RadonOpCodes::try_from(*integer as u8)
+                let [raw_op_code, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _] =
```
[the remark this answers is not in the document]
Comment (KiChjang) [reply rad/src/script.rs]: It just got stabilized: https://blog.rust-lang.org/2020/03/12/Rust-1.42.html, and I was planning to use this, but I wasn't sure what the project's minimum required rust version is, and I also don't know which rust version CI is using, so I opted for using underscores instead.

**Comment by tmpolaczyk** [comment]
Ok perfect! I think we will leave the two problematic lints enabled, and if they give us any problems in the future we can just disable them later.

Before merging, there are a few requirements on commit names.
We try to follow this commit convention: https://www.conventionalcommits.org/en/v1.0.0-beta.2/

So for example, `fix: disable dangerous casts`. You can rebase with `upstream/master` in interactive mode and use that to rename (and optionally squash) the commits.

    git rebase -i upstream/master

Another requirement is to sign all your commits. I can't find the guide we used to add a signing key to github, but once you do, you can use this magic command to sign all your commits:

   git rebase --exec 'git commit --amend --no-edit -n -S' -i upstream/master

This can be seen in github because the commits show a green "Verified" badge.

Once the PR is approved, you can merge it using

    git push upstream branch-name:master

Yes, all this should be explained in https://github.com/witnet/witnet-rust/blob/master/CONTRIBUTING.md, but I don't know why that's so outdated, sorry.

**Comment by KiChjang** [comment]
Alright, this PR should be ready to go, just waiting for an approval now.

**Comment by KiChjang** [comment]
@tmpolaczyk Unfortunately I don't have push access to the master branch, so I can't push to upstream master, nor can I merge this PR with the green button. Perhaps someone else will have to merge it?

**Comment by aesedepece** [comment]
@KiChjang You are invited to join write access:
https://github.com/witnet/witnet-rust/invitations

**Comment by tmpolaczyk** [comment]
Thank you!

**Comment by KiChjang** [comment]
Thanks for the write access!

**Comment by aesedepece** [comment]
@KiChjang welcome aboard!

**Pull request merged**